pyiron_base/project/delayed.py

```diff
@@ -0,0 +1,273 @@
+                ).values()
+            )
+        else:
+            output_dict[k] = v
+    return output_dict
+
+
+def draw(node_dict: dict, edge_lst: list):
+    """
+    Draw graph of nodes and edges
+
+    Args:
+        node_dict (dict): Dictionary of nodes
+        edge_lst (list): List of edges
+    """
+    import matplotlib.pyplot as plt
+    import networkx as nx
+    from IPython.display import SVG, display
+
+    graph = nx.DiGraph()
+    for k, v in node_dict.items():
+        graph.add_node(k, label=str(k).rsplit("_", 1)[0] + "=" + str(v))
+    for edge in edge_lst:
+        graph.add_edge(edge[1], edge[0])
+    svg = nx.nx_agraph.to_agraph(graph).draw(prog="dot", format="svg")
+    display(SVG(svg))
+    plt.show()
+
+
+class Selector:
+    def __init__(self, obj: object, selector: str):
+        self._obj = obj
+        self._selector = selector
+
+    def __getattr__(self, name: str):
+        if self._selector == "files" and name in self._obj._output_file_lst:
+            obj_copy = self._obj.__copy__()
+            obj_copy._output_file = name
+            return obj_copy
+        elif self._selector == "output" and name in self._obj._output_key_lst:
+            obj_copy = self._obj.__copy__()
+            obj_copy._output_key = name
+            return obj_copy
+        else:
+            raise AttributeError()
+
+
+class DelayedObject:
+    def __init__(
+        self,
+        function: callable,
+        *args,
+        output_key: Optional[str] = None,
+        output_file: Optional[str] = None,
+        output_file_lst: list = [],
+        output_key_lst: list = [],
+        list_length: Optional[int] = None,
+        list_index: Optional[int] = None,
+        **kwargs,
+    ):
+        self._input = {}
+        self._function = function
+        try:
+            self._input.update(
+                inspect.signature(self._function).bind(*args, **kwargs).arguments
+            )
+        except TypeError:
+            pass
+        self.__name__ = "DelayedObject"
+        self._result = None
+        self._output_key = output_key
+        self._output_file = output_file
+        self._output_key_lst = output_key_lst
+        self._output_file_lst = output_file_lst
+        self._list_length = list_length
+        self._list_index = list_index
+
+    def get_python_result(self):
+        return getattr(self._result.output, self._output_key)
+
+    def get_file_result(self):
+        return getattr(self._result.files, self._output_file)
+
+    def result(self):
+        if self._result is None:
+            input_dict = recursive_dict_resolve(input_dict=self._input)
+            if "kwargs" in input_dict.keys():
+                input_dict.update(input_dict.pop("kwargs"))
+            if "args" in input_dict.keys():
+                args = input_dict.pop("args")
+                self._result = self._function(*args, **input_dict)
+            else:
+                self._result = self._function(**input_dict)
```

Comment:
There are two steps done in one function here. I would suggest to separate `def result` and `def get_result` with

```python
    @cached_property
    def result(self):
        input_dict = recursive_dict_resolve(input_dict=self._input)
        if "kwargs" in input_dict.keys():
            input_dict.update(input_dict.pop("kwargs"))
        if "args" in input_dict.keys():
            args = input_dict.pop("args")
            return self._function(*args, **input_dict)
        else:
            return self._function(**input_dict)
```

and

```python
    def get_result(self):
        if self._output_key is not None:
            return self.get_python_result()
        elif self._output_file is not None:
            return self.get_file_result()
        elif self._list_index is not None:
            return self.result[self._list_index]
        else:
            return self.result
```

And everywhere `self._result` should be replaced by `self.result` and the current `self.result` should be replaced by `self.get_result()`. Not to forget to import `cached_property`

Reply:
maybe `result` should indeed be a private function. I leave the decision to you.

Reply:
I moved the evaluation of the function to a separate function.